lms/djangoapps/courseware/plugins.py

```diff
@@ -0,0 +1,153 @@
+from django.utils.translation import ugettext_noop as _
+from opaque_keys.edx.keys import CourseKey
+from xmodule.modulestore.django import modulestore
+
+from openedx.core.djangoapps.content.course_overviews.models import CourseOverview
+from openedx.core.djangoapps.course_apps.plugin import CourseApp
+from openedx.core.lib.courses import get_course_by_id
+
+User = get_user_model()
+
+TEXTBOOK_ENABLED = settings.FEATURES.get("ENABLE_TEXTBOOK", False)
+
+
+class ProgressCourseApp(CourseApp):
+    """
+    Course app config for progress app.
+    """
+
+    app_id = "progress"
+    name = _("Progress")
+    description = _("Allow students to track their progress throughout the course.")
+
+    @classmethod
+    def is_available(cls, course_key: CourseKey) -> bool:
+        """
+        The progress course app is always available.
+        """
+        return True
+
+    @classmethod
+    def is_enabled(cls, course_key: CourseKey) -> bool:
+        """
+        The progress course status is stored in the course module.
+        """
+        return not CourseOverview.get_from_id(course_key).hide_progress_tab
+
+    @classmethod
+    def set_enabled(cls, course_key: CourseKey, enabled: bool, user: 'User') -> bool:
+        """
+        The progress course enabled/disabled status is stored in the course module.
+        """
+        course = get_course_by_id(course_key)
+        course.hide_progress_tab = not enabled
+        modulestore().update_item(course, user.id)
+        return enabled
+
+    @classmethod
+    def get_allowed_operations(
+        cls, course_key: CourseKey, user: Optional[User] = None
+    ) -> Dict[str, bool]:  # pylint: disable=unused-argument
+        """
+        Returns the allowed operations for the app.
+        """
+        return {
+            "enable": True,
+            "configure": True,
+        }
+
+
+class TextbooksCourseApp(CourseApp):
+    """
+    Course app config for textbooks app.
+    """
+
+    app_id = "textbooks"
+    name = _("Textbooks")
+    description = _("Provide links to applicable resources for your course.")
+
+    @classmethod
+    def is_available(cls, course_key: CourseKey) -> bool:  # pylint: disable=unused-argument
+        """
+        The textbook app can be made available globally using a value in features.
+        """
+        return TEXTBOOK_ENABLED
+
+    @classmethod
+    def is_enabled(cls, course_key: CourseKey) -> bool:  # pylint: disable=unused-argument
+        """
+        Returns if the textbook app is globally enabled.
+        """
+        return TEXTBOOK_ENABLED
+
+    @classmethod
+    def set_enabled(cls, course_key: CourseKey, enabled: bool, user: 'User') -> bool:
+        """
+        The textbook app can be globally enabled/disabled.
+
+        Currently, it isn't possible to enable/disable this app on a per-course basis.
+        """
+        raise ValueError("The textbook app can not be enabled/disabled for a single course.")
```

Comment:
Why does this raise ValueError? Should it not return TEXTBOOK_ENABLED just like the WikiApp?

Reply:
I was asked to make the error explicit here: https://github.com/edx/edx-platform/pull/27542#discussion_r642893125